src/julia_gc.cc

```diff
@@ -1089,3 +1381,61 @@ void MarkJuliaWeakRef(void * p)
     if (JMarkTyped(p, jl_weakref_type))
         YoungRef++;
 }
+
+
+#ifdef TEST_JULIA_GC_INTERNALS
+
+struct IntCmp {
+    static int cmp(int a, int b)
+    {
+        return a - b;
+    }
+};
+
+__attribute__((constructor)) void TestScapegoatTree()
+{
+    const Int                    N = 1024 * 1024;
+    BalancedTree<int, IntCmp> * btree = BalancedTree<int, IntCmp>::make();
+    btree->init();
+    for (int i = 0; i < N; i++) {
+        btree->insert(i);
+    }
+    if (btree->count() != N)
+        abort();
+    if (height_to_size[btree->depth() - 1] > btree->count())
+        abort();
+    for (int i = 0; i < N; i++) {
+        if (!btree->find(i))
+            abort();
+    }
+    for (int i = 1; i < N; i++) {
+        btree->remove(i);
+    }
+    if (btree->count() != 1 || btree->depth() != 1)
+        abort();
+    btree->remove(0);
+    for (int i = 0, j = 0; i < N; i++, j = (5 * j + 1) & (N - 1)) {
```

Comment:
These tests (to me) don't seem complete enough -- they don't test that after each insertion and removal the tree is correct, they don't try removing a value from the tree which isn't present, or adding a value more than once. They also only seem to test 'find' returning true.

Reply:
Yeah, the tests are incomplete. At the moment, I wasn't sure whether using C++ was even acceptable, so I didn't want to invest too much time without being certain that I wasn't going to waste it.

Reply:
Okay, I've also added some array tests.